File: frontend/src/api/scheduleMeta.ts
import { apiClient } from './client'
import type { ScheduleStatus } from '../types'

export interface ScheduleSnapshot {
  id: number
  year: number
  month: number
  version_number: string
  name: string
  created_at: string
}

export interface ChangeLogEntry {
  id: number
  year: number
  month: number
  employee_id: number
  day: number
  old_shift: string
  new_shift: string
  reason: string | null
  created_at: string
}

export interface SnapshotDiffChange {
  employee_id: number
  employee_name: string | null
  day: number
  old_shift: string | null
  new_shift: string | null
}

export async function getScheduleStatus(
  year: number,
  month: number,
): Promise<{ year: number; month: number; status: ScheduleStatus }> {
  const { data } = await apiClient.get(`/schedule-status/${year}/${month}`)
  return data
}

export async function setScheduleStatus(
  year: number,
  month: number,
  status: ScheduleStatus,
): Promise<{ year: number; month: number; status: ScheduleStatus }> {
  const { data } = await apiClient.put(`/schedule-status/${year}/${month}`, { status })
  return data
}

export async function getSnapshots(year: number, month: number): Promise<ScheduleSnapshot[]> {
  const { data } = await apiClient.get<ScheduleSnapshot[]>(`/snapshots/${year}/${month}`)
  return data
}

export async function createSnapshot(
  year: number,
  month: number,
  name?: string,
): Promise<ScheduleSnapshot> {
  const { data } = await apiClient.post<ScheduleSnapshot>(`/snapshots/${year}/${month}`, { name })
  return data
}

export async function restoreSnapshot(id: number): Promise<ScheduleSnapshot> {
  const { data } = await apiClient.post<ScheduleSnapshot>(`/snapshots/${id}/restore`)
  return data
}

export async function diffSnapshots(
  aId: number,
  bId: number,
): Promise<{ changes: SnapshotDiffChange[] }> {
  const { data } = await apiClient.get(`/snapshots/diff/${aId}/${bId}`)
  return data
}

export async function getChangeLogs(year: number, month: number): Promise<ChangeLogEntry[]> {
  const { data } = await apiClient.get<ChangeLogEntry[]>(`/change-logs/${year}/${month}`)
  return data
}

File: backend/app/api/routes/schedule_meta.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.deps import get_db
from app.schemas.schedule_meta import (
    ChangeLogOut,
    SnapshotCreateRequest,
    SnapshotOut,
    StatusUpdateRequest,
)
from app.services import change_log_service, snapshot_service, status_service

router = APIRouter()


@router.get("/schedule-status/{year}/{month}")
def get_schedule_status(year: int, month: int, db: Session = Depends(get_db)):
    return {
        "year": year,
        "month": month,
        "status": status_service.get_status(db, year, month),
    }


@router.put("/schedule-status/{year}/{month}")
def set_schedule_status(
    year: int, month: int, req: StatusUpdateRequest, db: Session = Depends(get_db)
):
    try:
        if req.status == "published":
            status_service.publish(db, year, month)
        else:
            status_service.set_status(db, year, month, req.status)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(exc))
    return {"year": year, "month": month, "status": req.status}


@router.get("/snapshots/{year}/{month}", response_model=list[SnapshotOut])
def list_snapshots(year: int, month: int, db: Session = Depends(get_db)):
    return snapshot_service.list_snapshots(db, year, month)


@router.post(
    "/snapshots/{year}/{month}",
    response_model=SnapshotOut,
    status_code=status.HTTP_201_CREATED,
)
def create_snapshot(
    year: int, month: int, req: SnapshotCreateRequest, db: Session = Depends(get_db)
):
    return snapshot_service.create_snapshot(db, year, month, name=req.name)


@router.post("/snapshots/{snapshot_id}/restore", response_model=SnapshotOut)
def restore_snapshot(snapshot_id: int, db: Session = Depends(get_db)):
    try:
        return snapshot_service.restore_snapshot(db, snapshot_id)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


@router.get("/snapshots/diff/{snapshot_id_a}/{snapshot_id_b}")
def diff_snapshots(
    snapshot_id_a: int, snapshot_id_b: int, db: Session = Depends(get_db)
):
    try:
        return snapshot_service.diff_snapshots(db, snapshot_id_a, snapshot_id_b)
    except ValueError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc))


@router.get("/change-logs/{year}/{month}", response_model=list[ChangeLogOut])
def list_change_logs(year: int, month: int, db: Session = Depends(get_db)):
    return change_log_service.list_change_logs(db, year, month)
